ckanext/datagovuk/upload.py

```
@@ -0,0 +1,149 @@
+
+    # Init connection to S3
+    bucket = setup_s3_bucket()
+
+    # Get content type and extension
+    content_type, _ = mimetypes.guess_type(
+        resource.get('url', ''))
+    extension = mimetypes.guess_extension(content_type)
+
+    # Upload to S3
+    timestamp = datetime.datetime.utcnow()
+    pkg = toolkit.get_action('package_show')(context, {'id': resource['package_id']})
+    s3_filepath = (pkg.get('name')
+                   + '/'
+                   + 'resources'
+                   + '/'
+                   + resource.get('timestamp', timestamp.strftime("%Y-%m-%dT%H-%M-%SZ"))
+                   + '-'
+                   + slugify(resource.get('name'), to_lower=True)
+                   + extension)
+
+    # If file is currently being uploaded, the file is in resource['upload']
+    if isinstance(resource.get('upload', None), cgi.FieldStorage):
+        logger.info("File is being uploaded")
+        resource['upload'].file.seek(0)
+        body = resource['upload'].file
+    # If resource.get('url_type') == 'upload' then the resource is in CKAN file system
+    elif resource.get('url_type') == 'upload':
+        logger.info("File is on CKAN file store")
+        upload = uploader.ResourceUpload(resource)
+        filepath = upload.get_path(resource['id'])
+        try:
+            body = open(filepath, 'r')
+        except OSError:
+            abort(404, _('Resource data not found'))
+    else:
+        return ## in datagovuk, we don't want to upload from URL
+
+    try:
+        logger.info("Uploading resource %s to S3" % resource.get('name', ''))
+        bucket.Object(s3_filepath).delete()
+        obj = bucket.put_object(Key=s3_filepath,
+                                Body=body,
+                                ContentType=content_type)
+        obj.Acl().put(ACL='public-read')
+        logger.info("Successfully uploaded resource %s to S3" % resource.get('name', ''))
+
+    except Exception as exception:
```

We should catch a more specific S3 error rather than a catch all exception